tools/profile_endpoints.py

```diff
@@ -0,0 +1,315 @@
+
+
+# Endpoint Profiling Functions
+def profile_get_snapshot_builder_count(token):
+    """
+    Profiles the getSnapshotBuilderCount endpoint and returns the time taken.
+
+    Args:
+        token (str): The access token for authentication.
+
+    Returns:
+        str: A string indicating the average time taken for the request.
+    """
+    print("Profiling getSnapshotBuilderCount endpoint")
+
+    # Define the bodies for the POST request
+    bodys = [
+        {
+            "cohorts": [
+                {
+                    "name": "",
+                    "criteriaGroups": [
+                        {
+                            "name": "Group 1",
+                            "mustMeet": True,
+                            "meetAll": False,
+                            "criteria": [
+                                {"kind": "range", "id": 0, "low": 1955, "high": 2021},
+                                {"kind": "list", "id": 2, "values": [38003563]},
+                                {"kind": "list", "id": 1, "values": [8507]},
+                            ],
+                        }
+                    ],
+                }
+            ]
+        },
+        {
+            "cohorts": [
+                {
+                    "name": "",
+                    "criteriaGroups": [
+                        {
+                            "name": "Group 1",
+                            "mustMeet": True,
+                            "meetAll": False,
+                            "criteria": [
+                                {"kind": "domain", "id": 19, "conceptId": 4042140},
+                                {"kind": "domain", "id": 19, "conceptId": 441840},
+                                {"kind": "domain", "id": 27, "conceptId": 46234754},
+                                {"kind": "list", "id": 3, "values": [8515]},
+                            ],
+                        }
+                    ],
+                }
+            ]
+        },
+        {
+            "cohorts": [
+                {
+                    "name": "",
+                    "criteriaGroups": [
+                        {
+                            "name": "Group 1",
+                            "mustMeet": True,
+                            "meetAll": False,
+                            "criteria": [
+                                {"kind": "range", "id": 0, "low": 1992, "high": 2021},
+                                {"kind": "list", "id": 1, "values": [0]},
+                                {"kind": "list", "id": 2, "values": [38003563]},
+                                {"kind": "list", "id": 3, "values": [8527]},
+                            ],
+                        }
+                    ],
+                }
+            ]
+        },
+    ]
+
+    time_length = 0
+    for body in bodys:
+        time_taken = make_post_request("count", token, body)
+        time_length += time_taken
+
+    return f"Time taken: {time_length / len(bodys)} seconds"
+
```

Comment:
It is tempting to try and average the times of multiple calls to get an 'average' time for an operation, but we know that, in bigquery at least, some caching is done in the server, so it returns a cached response for a duplicate query. If we wanted to avoid getting misled by caching, we could try a different approach where the benchmarking code randomizes part of a query to try and avoid cache hits.

Reply:
I haven't tried querying with big query but I'm not sure if azure caches these values because I found there is no performance increases on hitting the same query several times. But, I do agree that randomizing a query may help benchmarking. I'll look into ways of randomizing this!

Reply:
I think it's OK to push it as-is for now. The data for azure is good at least, and that's where we need to focus right now anyway.